File: terminal-games-sdk/src/audio/mixer.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

use super::{FRAME_SIZE, Instance, host};
use std::sync::{Arc, Mutex, OnceLock, Weak};

static GLOBAL_MIXER: OnceLock<Mixer> = OnceLock::new();

pub fn mixer() -> &'static Mixer {
    GLOBAL_MIXER.get_or_init(Mixer::new)
}

struct MixerState {
    instances: Vec<Weak<Instance>>,
    master_volume: f32,
    mix_buffer: Vec<f32>,
    scratch_buffer: Vec<f32>,
}

pub struct Mixer {
    state: Mutex<MixerState>,
}

impl Mixer {
    fn new() -> Self {
        Self {
            state: Mutex::new(MixerState {
                instances: Vec::new(),
                master_volume: 1.0,
                mix_buffer: Vec::new(),
                scratch_buffer: Vec::new(),
            }),
        }
    }

    /// Sets the master volume for all mixed audio.
    ///
    /// 0.0 = silent, 1.0 = full volume.
    pub fn set_master_volume(&self, volume: f32) {
        let mut state = self.state.lock().unwrap();
        state.master_volume = volume;
    }

    pub fn master_volume(&self) -> f32 {
        let state = self.state.lock().unwrap();
        state.master_volume
    }

    pub(crate) fn add_instance(&self, instance: &Arc<Instance>) {
        let mut state = self.state.lock().unwrap();
        state.instances.push(Arc::downgrade(instance));
    }

    fn mix(&self, num_frames: usize) -> Vec<f32> {
        let mut state = self.state.lock().unwrap();

        let num_values = num_frames * super::CHANNELS;
        if state.mix_buffer.len() < num_values {
            state.mix_buffer.resize(num_values, 0.0);
            state.scratch_buffer.resize(num_values, 0.0);
        }

        for sample in &mut state.mix_buffer[..num_values] {
            *sample = 0.0;
        }

        let live_instances: Vec<_> = state
            .instances
            .iter()
            .filter_map(|weak| weak.upgrade())
            .collect();
        state.instances.retain(|weak| weak.strong_count() > 0);

        for instance in live_instances {
            for sample in &mut state.scratch_buffer[..num_values] {
                *sample = 0.0;
            }

            instance.fill_buffer(&mut state.scratch_buffer[..num_values]);

            for i in 0..num_values {
                state.mix_buffer[i] += state.scratch_buffer[i];
            }
        }

        let master_volume = state.master_volume;
        for sample in &mut state.mix_buffer[..num_values] {
            *sample *= master_volume;
            *sample = sample.clamp(-1.0, 1.0);
        }

        state.mix_buffer[..num_values].to_vec()
    }

    /// Performs one tick of mixing and writing to the host.
    ///
    /// This should be called periodically from the main loop, typically
    /// every 5-10ms for smooth audio.
    pub fn tick(&self) {
        let Some(audio_info) = host::info() else {
            return;
        };

        let target_buffer = FRAME_SIZE as u32 * 2;

        if audio_info.buffer_available >= target_buffer {
            return;
        }

        let needed = (target_buffer - audio_info.buffer_available) as usize;
        let frames = (needed + FRAME_SIZE - 1) / FRAME_SIZE;
        let num_frames = frames * FRAME_SIZE;

        if num_frames == 0 {
            return;
        }

        let mixed = self.mix(num_frames);
        host::write(&mixed);
    }
}

File: pkg/audio/mixer.go
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

package audio

import (
	"sync"
	"time"
)

type Mixer struct {
	instances []*Instance
	mu        sync.Mutex
	running   bool
	volume    float32

	mixBuffer     []float32
	scratchBuffer []float32
}

var globalMixer = &Mixer{
	instances: make([]*Instance, 0),
	volume:    1.0,
}

func GetMixer() *Mixer {
	return globalMixer
}

// SetMasterVolume sets the master volume for all mixed audio.
//
// 0.0 = silent, 1.0 = full volume.
func (m *Mixer) SetMasterVolume(v float32) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.volume = v
}

func (m *Mixer) MasterVolume() float32 {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.volume
}

func (m *Mixer) addInstance(inst *Instance) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.instances = append(m.instances, inst)
}

func (m *Mixer) removeInstance(inst *Instance) {
	m.mu.Lock()
	defer m.mu.Unlock()
	for i, existing := range m.instances {
		if existing == inst {
			m.instances = append(m.instances[:i], m.instances[i+1:]...)
			break
		}
	}
}

func (m *Mixer) mix(numFrames int) []float32 {
	m.mu.Lock()
	defer m.mu.Unlock()

	numValues := numFrames * Channels
	if len(m.mixBuffer) < numValues {
		m.mixBuffer = make([]float32, numValues)
		m.scratchBuffer = make([]float32, numValues)
	}

	for i := range numValues {
		m.mixBuffer[i] = 0
	}

	for _, inst := range m.instances {
		if !inst.playing {
			continue
		}

		for i := range numValues {
			m.scratchBuffer[i] = 0
		}

		inst.fillBuffer(m.scratchBuffer[:numValues])

		for i := range numValues {
			m.mixBuffer[i] += m.scratchBuffer[i]
		}
	}

	for i := range numValues {
		sample := m.mixBuffer[i] * m.volume
		if sample > 1.0 {
			sample = 1.0
		} else if sample < -1.0 {
			sample = -1.0
		}
		m.mixBuffer[i] = sample
	}

	return m.mixBuffer[:numValues]
}

func startMixer() {
	globalMixer.mu.Lock()
	if globalMixer.running {
		globalMixer.mu.Unlock()
		return
	}
	globalMixer.running = true
	globalMixer.mu.Unlock()

	go func() {
		writer := NewAudioWriter(FrameSize * 4)

		for {
			neededFrames := writer.ShouldWrite()
			if neededFrames > 0 {
				mixed := globalMixer.mix(neededFrames)
				written := Write(mixed)
				if written > 0 {
					writer.NextPTS += uint64(written)
				}
			}

			time.Sleep(5 * time.Millisecond)
		}
	}()
}
